### PR TITLE
implement tail in go

### LINKED ISSUE #2: Implement tail(1) from scratch in Go
... so that `go run . < in.txt` does the same as `tail < in.txt`

### PR DESCRIPTION
**Pull request by justinmueller05082000**
fixes #2

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	"fmt"
+	"io"
+	"os"
+)
+
+func main() {
+	readAndPrintFile()
+}
+
+func readAndPrintFile() {
```
Comment (Al2Klimov) [inline main.go]: What's this for?

File: main.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	if lineCounter < 10 {
+		lineCounter = 0
+	} else {
+		lineCounter = lineCounter - 10
+	}
+
+	for i := lineCounter; i < len(allData); i++ {
+		fmt.Printf("%s", allData[i])
```
Comment (Al2Klimov) [inline main.go]: How did you output byte arrays like this in previous projects?

File: main.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	"os"
+)
+
+func main() {
+	readAndPrintFile()
+}
+
+func readAndPrintFile() {
+	buf := bufio.NewReader(os.Stdin)
+	var lineCounter int
```
Comment (Al2Klimov) [inline main.go]: What's `lineCounter` for?

File: main.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+	if lineCounter < 10 {
+		lineCounter = 0
+	} else {
+		lineCounter = lineCounter - 10
+	}
+
+	for i := lineCounter; i < len(allData); i++ {
+		fmt.Printf("%s", allData[i])
+	}
+}
```
Comment (Al2Klimov) [inline main.go]: Get rid of that red thing.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: main.go (hunk 3 above)
Comment (justinmueller05082000) [reply main.go]: to count the lines of the file but if the end of the file is an empty line the program does not count it as a line

Comment (Al2Klimov) [reply main.go]: What about just using `allData`?

Comment (justinmueller05082000) [reply main.go]: did not work properly thats why I use lineCounter

File: main.go (hunk 4 above)
Comment (justinmueller05082000) [reply main.go]: the for loop?

File: main.go (hunk 2 above)
Comment (justinmueller05082000) [reply main.go]: os.Stdout.Write?

File: main.go (hunk 4 above)
Comment (Al2Klimov) [reply main.go]: Click on "Files changed" and scroll down.

File: main.go (hunk 3 above)
Comment (Al2Klimov) [reply main.go]: What exactly "did not work properly"?

File: main.go (hunk 2 above)
Comment (Al2Klimov) [reply main.go]: Exactly!

File: main.go (hunk 3 above)
Comment (justinmueller05082000) [reply main.go]: the len of dataAll is 4 if i have 3 lines + 1 empty line at the end but tail does not count an empty line at the end as a line so tail just ignores it and prints 10 lines upwards from the last-not-empty-line in the file

Comment (Al2Klimov) [reply main.go]: Why don't you just throw the empty line away?

File: main.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+		}
+		allData = append(allData, data)
+		if dataErr == io.EOF {
+			if bytes.Compare(data, []byte("")) == 0 {
+			}
+			break
+		}
+	}
+
+	if bytes.Compare(allData[len(allData)-1], []byte("")) == 0 {
```
Comment (Al2Klimov) [inline main.go]: Any way to shorten `bytes.Compare(..., []byte("")) == 0`?

File: main.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+		allData = allData[:(len(allData) - 1)]
```
Comment (Al2Klimov) [inline main.go]: Can you remove some of the ( ) here?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: main.go (hunk 5 above)
Comment (justinmueller05082000) [reply main.go]: i dont think so

Comment (Al2Klimov) [reply main.go]: How long is a `[]byte("")`?

Comment (justinmueller05082000) [reply main.go]: 0?

Comment (Al2Klimov) [reply main.go]: Exactly.

File: main.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+	for {
+		data, dataErr := buf.ReadBytes('\n')
+		if dataErr != nil && dataErr != io.EOF {
+			fmt.Fprintln(os.Stderr, dataErr)
+			os.Exit(1)
+		}
+		allData = append(allData, data)
+		if dataErr == io.EOF {
+			if bytes.Compare(data, []byte("")) == 0 {
```
Comment (Al2Klimov) [inline main.go]: And here?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: main.go (hunk 7 above)
Comment (justinmueller05082000) [reply main.go]: forgot to remove that

File: main.go [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+		if len(allData) > 10 && len(data) != 0 {
+			allData = allData[1:]
+		}
+
+		if dataErr == io.EOF {
+			break
+		}
+	}
+	if len(allData[len(allData)-1]) == 0 {
```
Comment (Al2Klimov) [inline main.go]: Separate the for and if by an empty line.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
* [x] `git reset --soft HEAD~$(git rev-list --count $(git merge-base HEAD master)..HEAD^)`
* [x] `git commit --amend --no-edit`
* [x] `git push -f origin feature/implement-tail-in-go-2`

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Write a test protocol.

**Comment by justinmueller05082000** [comment]
**Test protocol:**

**in.txt:**
```
one
two
three
four
five
six
seven
eight
nine
ten
eleven
twelve
thirteen
fourteen
```

**Case 1:**
**Terminal:** `go run . < in.txt`
**Output:**
```
five
six
seven
eight
nine
ten
eleven
twelve
thirteen
fourteen
```
**Terminal:** `tail < in.txt`
**Output:**
```
five
six
seven
eight
nine
ten
eleven
twelve
thirteen
fourteen
```

**Conclusion:**
Both outputs are equal.

**Review by Al2Klimov — APPROVED** [review]

**Pull request merged**